feat: integrate zkLogin

olehmisar opened the pull request:
Uses `@shield-labs/zklogin` package.

TODO:
- [x] Backup and recover methods in the solidity contract
- [x] Backup and recover boilerplate
- [x] Fix type errors
- [x] Grab JWT from Google
- [ ] Trustless posting of Google public keys on chain
- [ ] Service to maintain account_id salt to preserve user privacy (currently hardcoded to be 0)
- [x] Fix `experimental_addBackup`
- [x] Implement `experimental_recover`
- [ ] Implement `experimental_removeBackup`, `experimental_listBackups`
- [ ] Use UltraHonk for 10-second proving times (with higher gas cost). Blocked by https://github.com/olehmisar/zklogin/pull/5

The spec should look something like this:
```ts
{
  "method": "experimental_addBackup",
  "params": [{
    "address": "0x...",       // Optional: Defaults to the connected account
    "backupOptions": [{        // Required for "add" and "remove" actions
      "type": "openId" | "email" | "socialRecovery",
      "provider": "google" | "apple" | "email" | "custom",
      "identifier": "user@example.com", // Email or unique identifier
    }]
  }]
}

{
  "method": "experimental_recover",
  "params": [{
    "address": "0x...",
    "backupOption": {
      "type": "openId" | "email" | "socialRecovery",
      "provider": "google" | "apple" | "email" | "custom",
      "proof": "..."       // Proof of ownership
    },
    "newAuthentication": {
      "passkey": "..."     // Details for setting up a new passkey or authentication method
    }
  }]
}
```

olehmisar commented:
zkLogin requires oauth setup on the server. Playground does not have a server because it's a vite app.

The server is needed because Google will send OpenID token to e.g., `porto.ithaca.xyz/auth/callback` and the server must send the token to the frontend.

tmm commented:
With the new [Vite Environment API](https://vite.dev/guide/api-environment), we should be able to spin up an API server in parallel at some point.

For now, you want to create a new `api` directory inside playground that uses [Hono](https://hono.dev/) (`pnpm create hono`) or use an API route in [`app`](https://github.com/ithacaxyz/porto/tree/main/app/src/app).

olehmisar commented:
I implemented `addBackup` and `recover` RPC methods and also added simple playground UIs. Still missing items are: `removeBackup` and `listBackups` but it shouldn't be a blocker for a review. 

We probably won't be going with UltraHonk because, per my conversation with Georgios, users will rather wait few more seconds (15s vs 30s) than pay twice as much for the gas fee (500k gas vs 1mil gas).

jxom commented:
It looks like some bundlers are also not happy with usage of pedersen in `@aztec/foundation` as it also includes `node:crypto` which doesn't exist in the browser – this is a side-effect of using the `Fr` class (I don't think this is tracked in https://github.com/AztecProtocol/aztec-packages/issues/8881). I am wondering if your zklogin package could instead perhaps use @paulmillr's `pederson` hash fn from `scure-starknet`: https://github.com/paulmillr/scure-starknet/blob/main/index.ts#L253-L258 instead of `@aztec/foundation`? This may remove all the target & resolution config in `vite.config.ts`, as well as the `Buffer` polyfill too.

olehmisar commented:
@jxom i would instead remove @aztec/foundation altogether in favour of @aztec/bb.js. It will still require polyfilling `Buffer` at the moment but it's much easier to remove `Buffer` from bb.js than from foundation. I can work with Aztec team on removing `Buffer` from bb.js

jxom commented:
Sounds great! I am also wondering what we can do to mitigate the massive bundle size jump. It seems that this adds nearly ~890kB to the **gzip minified** bundle.

```
$ pnpm check:size

  Package size limit has exceeded by 890.87 kB
  Size limit:   54 kB
  Size:         944.87 kB with all dependencies, minified and brotlied
  Loading time: 18.5 s    on slow 3G
  Running time: 2.6 s     on Snapdragon 410
  Total time:   21 s
```

olehmisar commented:
@jxom I removed @aztec/foundation and all polyfills. `target: "next"` are still required due to https://github.com/AztecProtocol/aztec-packages/issues/10184.

The large size of bundle is due to the size of `bb.js` wasm. It is loaded lazily using `await import`. I can consult with Aztec team what can be done to reduce the size.

olehmisar commented on src/internal/accountDelegation.ts:
```diff
@@ -606,6 +608,71 @@ export declare namespace prepareInitialize {
   }
 }
 
+const zkLogin = new zklogin.ZkLogin()
```
probably shouldn't be initialized here

olehmisar commented on src/internal/rpcSchema.ts:
```diff
@@ -144,3 +159,35 @@ export type PrepareImportAccountReturnType = {
+  proof: NonNullable<Awaited<ReturnType<zklogin.ZkLogin['proveJwt']>>>
+}
+
+// TODO: import from `@shield-labs/zklogin`
+type ZkLoginProvider = 'google' // | 'apple'
```
fix this todo

olehmisar reviewed (commented):
notes

gakonst commented:
What should we do about the published certs? Maybe we have versions in the registry and users can opt-in to a rotation?

olehmisar commented:
@gakonst certs are rotated regularly, so we can’t make them immutable. We can probably leverage a blind oracle (aka TLSNotary) to publish certs. It’s an MPC network running inside a bunch of TEEs that do not see what they signs. It requires Porto to depend on an oracle service provider. vlayer.xyz provides this kind of service.

gakonst commented:
not comfortable w/ introducing such 3rd party deps...will think about it. maybe we run this ourselves.

olehmisar commented:
the choice boils down to a multisig oracle.

TLDR: MPC with no secret data = a multisig. A blind oracle that posts exclusively Google/Apple certs = a multisig.

Why other options will not work:
1. MPC does not make sense because there is no secret data, so MPC would be a pure overhead with no advantage over a multisig
2. Blind oracles (aka TLSNotary, aka zkTLS, aka WebProofs) - similar to the option 1. If we using a blind oracle to post ONLY google/apple certs, then we don't benefit from the blindness property. A blind oracle is only useful when it posts different kinds of data. It's similar to an anonymity set in a privacy pool - the larger the anonymity set the more private the pool is. In blind oracles, the more types of data the oracle posts, the harder it is for the oracle to tamper with data because it won't know what it is tampering with: a Google cert or e.g. an Uber proof of driver.

gakonst commented:
>  I can consult with Aztec team what can be done to reduce the size.

Circling back...  @olehmisar did u get any news on this? That'd be epic if we could get done. I'm talking with the CF team about potentially doing SXG for the certs, circling back soon.

olehmisar commented:
@gakonst yes, we can reduce the size at least by a factor of x2. And if that’s not enough, we can then bundle an application-specific wasm that only includes functions needed for zklogin to reduce size even more

gakonst commented:
Also worth looking into https://github.com/zkemail/jwt-tx-builder

olehmisar commented:
They are similar. ZkLogin has a nicer API imo as I tried to keep it as simple as possible. Are you missing anything specific in zkLogin from jwt-tx-builder?

Also, @gakonst are there any updates from cloudflare regarding jwt public keys over SXG?

jxom commented:
Closing as the implementation is well stale. We will revisit at a later date.
